### PR TITLE
Sync sent mailboxes after sending an email

### PR DESCRIPTION
**Pull request by StCyr**
Fix https://github.com/nextcloud/mail/issues/3598

Maybe there's some async things to improve though

Signed-off-by: Cyrille Bollu <cyrpub@bollu.be>

### CODE REVIEW

**Comment by StCyr** [comment]
A fix to https://github.com/nextcloud/mail/issues/3598 becomes more and more important with the introduction of threading (sent emails must be sync'ed locally before they can be part of a thread)

**Comment by ChristophWurst** [comment]
1) This logic should be done in an action, like we have one to sync all inboxes
2) You can't sync a mailbox if it's not cached and fetched yet. You need to do this before IIRC.
3) It would be a lot more efficient to do the sync on the back-end with sync criterion *new* only. Sync'ing for changed messages is typically expensive and thus slower. For the thread we just need the new message, tho.

**Comment by StCyr** [comment]
>     1. This logic should be done in an action, like we have one to sync all inboxes

Will do that 
 
>     2. You can't sync a mailbox if it's not cached and fetched yet. You need to do this before IIRC.

Hmmm, isn't caching something that happens only once (the first time we open a folder)? 

Just for my personal knowledge: Not that I think I shouldn't handle this case 

>     3. It would be a lot more efficient to do the sync on the back-end with sync criterion _new_ only. Sync'ing for changed messages is typically expensive and thus slower. For the thread we just need the new message, tho.

I have implemented this. do you like it?

File: src/components/Envelope.vue [hunk 1]
```diff
@@ -171,13 +171,14 @@ export default {
 				const recipients = [this.data.to, this.data.cc].flat().map(function(recipient) {
 					return recipient.email
 				})
-				return recipients.length > 0 ? recipients[0] : undefined
+				return recipients.length > 0 ? recipients[0] : 'Blind copy recipients only'
```
Comment (ChristophWurst) [inline src/components/Envelope.vue:174]: :-1: I'm fine with this enhancement but 1) it's not translated and 2) this is unrelated to the topic of this PR. Please revert and submit separately :v:

File: src/components/Envelope.vue [hunk 2]
```diff
@@ -171,13 +171,14 @@ export default {
...
 			}
 
 			// Show sender avatar in other mailbox types
 			if (this.data.from.length > 0) {
 				return this.data.from[0].email
 			} else {
+				console.error('No email found in from field', this.data)
```
Comment (ChristophWurst) [inline src/components/Envelope.vue:181]: same

File: src/components/NewMessageDetail.vue [hunk 3]
```diff
@@ -144,12 +144,18 @@ export default {
 				logger.debug('detected navigation to current (new) draft UID, not reloading')
 				return
 			}
-			logger.debug('the  draft ID changed, we have to fetch the draft', {
-				currentId: this.draft.databaseId,
-				newId: to.params.draftId,
-			})
 
-			this.fetchMessage()
+			if (to.params.threadId === 'new') {
```
Comment (ChristophWurst) [inline src/components/NewMessageDetail.vue:148]: another unrelated change, right?

File: src/store/actions.js [hunk 4]
```diff
@@ -406,7 +406,18 @@ export default {
 			return envelopes
 		})
 	},
-	syncEnvelopes({ commit, getters, dispatch }, { mailboxId, query, init = false }) {
+	syncEnvelopesByRole({ commit, getters, dispatch }, { role, accountId, query, init = false, criteria = 'all' }) {
+		const mailboxes = getters.getMailboxes(accountId)
```
Comment (ChristophWurst) [inline src/store/actions.js:410]:
```suggestion
		getters.getMailboxes(accountId)
		    .filter(mb => mb.specialRole === role)
		    .map(mb => dispatch(...))
```

for short and concise code :wink:

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

File: src/components/NewMessageDetail.vue (hunk 3 above)
Comment (StCyr) [reply src/components/NewMessageDetail.vue:148]: yes, I'll create a new PR

File: src/components/Envelope.vue (hunk 2 above)
Comment (StCyr) [reply src/components/Envelope.vue:181]: ok

File: src/components/Envelope.vue (hunk 1 above)
Comment (StCyr) [reply src/components/Envelope.vue:174]: ok

File: lib/Controller/MailboxesController.php [hunk 5]
```diff
@@ -147,20 +147,26 @@ public function patch(int $id,
 	 *
 	 * @param bool $init
 	 * @param string|null $query
+	 * @param string|'all' $criteria
```
[the remark this answers is not in the document]
Comment (StCyr) [reply lib/Controller/MailboxesController.php:150]: I understand.

But having 3 booleans, would make the route very long:

```diff
diff --git a/appinfo/routes.php b/appinfo/routes.php
index 4f669ad09..068db5dea 100644
--- a/appinfo/routes.php
+++ b/appinfo/routes.php
@@ -91,8 +91,11 @@ return [
                ],
                [
                        'name' => 'mailboxes#sync',
-                       'url' => '/api/mailboxes/{id}/sync',
-                       'verb' => 'POST'
+                       'url' => '/api/mailboxes/{id}/sync/{criteria}',
+                       'verb' => 'POST',
+                       'defaults' => [
+                               'criteria' => 'all'
+                       ]
                ],
                [
                        'name' => 'mailboxes#clearCache',
```

We could maybe use an int that just copies Horde's implementation `Horde_Imap_Client::SYNC_NEWMSGSUIDS | Horde_Imap_Client::SYNC_FLAGSUIDS | Horde_Imap_Client::SYNC_VANISHEDUIDS`?

[the remark this answers is not in the document]
Comment (ChristophWurst) [reply lib/Controller/MailboxesController.php:150]: I wouldn't include the booleans in the route but send them as GET parameters like `.../sync?syncNew=true`

**Review by ChristophWurst — APPROVED** [review]
Let's use https://github.com/nextcloud/mail/pull/3820 as discussed at the conf

**Review by ChristophWurst — DISMISSED** [review]
Wrong button :speak_no_evil:

**Comment by ChristophWurst** [comment]
> Let's use #3820 as discussed at the conf

Please update your branch. You can now do smart sync with only `account.sentMailboxId` :)

File: lib/Controller/MailboxesController.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChristophWurst) [reply lib/Controller/MailboxesController.php:150]: Can we do this? If you're busy I can take over. Just let me know :pray:

[the remark this answers is not in the document]
Comment (StCyr) [reply lib/Controller/MailboxesController.php:150]: right now I'm busy, playing with attachments. So, feel free to take this PR over if you like. 

Otherwise, I'll probably be able to go back at it in the next coming days.

**Comment by ChristophWurst** [comment]
@StCyr do you want to revive this branch and adapt to  #3820 or start a new one?

**Review by GretaD — APPROVED** [review]
tested and it works. didnt check the code

**Review by ChristophWurst — CHANGES_REQUESTED** [review]
The rest looks great! Thanks a lot!

File: lib/Controller/MailboxesController.php [hunk 6]
```diff
@@ -147,20 +147,33 @@ public function patch(int $id,
...
 	 * @throws ServiceException
 	 */
-	public function sync(int $id, array $ids = [], bool $init = false, string $query = null): JSONResponse {
+	public function sync(int $id, array $ids = [], bool $init = false, string $query = null, array $filter = []): JSONResponse {
 		$mailbox = $this->mailManager->getMailbox($this->currentUserId, $id);
 		$account = $this->accountService->find($this->currentUserId, $mailbox->getAccountId());
 
+		// What types of messages do we need to sync?
+		$toSync = 0;
+		if (!array_key_exists('syncChanged', $filter) or $filter['syncChanged'] === true) {
```
[the remark this answers is not in the document]
Comment (StCyr) [reply lib/Controller/MailboxesController.php:162]: yes true

File: src/components/Composer.vue [hunk 7]
```diff
@@ -629,13 +629,26 @@ export default {
...
+						mailboxId: account.sentMailboxId,
+						query: '',
+						init: false,
+						filter: {
+							syncChanged: false,
+							syncNew: true,
+							syncVanished: false,
+						},
+					})
+				}, 2000))
```
[the remark this answers is not in the document]
Comment (StCyr) [reply src/components/Composer.vue:651]: If I remember well, I'd noticed that sync'ing right after sending the email didn't work: The just-sent message wouldn't be fetched, like if I wasn't giving the IMAP server the time to update the mailbox. 

So, I added a timeout and it looked to me that 2 seconds was a wise choice, not too fast, not too slow.

Note that we might want to remove this sync once https://github.com/nextcloud/mail/pull/4161 is implemented though.

[the remark this answers is not in the document]
Comment (ChristophWurst) [reply src/components/Composer.vue:651]: > Note that we might want to remove this sync once #4161 is implemented though.

Good point. I guess if we have #4161 the only case where we need to trigger a sync after sending a message is when the user currently is in the *sent* mailbox, right? In any other case we don't need this special logic.

[the remark this answers is not in the document]
Comment (StCyr) [reply src/components/Composer.vue:651]: Right, I'll update accordingly whatever PR comes second 👍

**Comment by ChristophWurst** [comment]
I merged  #4161 because that solves more general problems with outdated mailboxes.

Now I think we don't even need the conditional filtered sync anymore. The only logic left is that when a message is sent and the currently open mailbox is the account's sent mailbox, we trigger an ordinary sync. @StCyr does that make senes?

**Comment by StCyr** [comment]
Ok, I've created a new PR => https://github.com/nextcloud/mail/pull/4206
